test/functional/wallet_orphanedreward.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test orphaned block rewards in the wallet."""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+class OrphanedBlockRewardTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        # Generate some blocks and obtain some coins on node 0.  We send
+        # some balance to node 1, which will hold it as a single coin.
+        self.nodes[0].generate(150)
+        self.nodes[0].sendtoaddress(self.nodes[1].getnewaddress(), 10)
+        self.nodes[0].generate(1)
+
+        # Get a block reward with node 1 and remember the block so we can orphan
+        # it later.
+        self.sync_blocks()
+        blk = self.nodes[1].generate(1)[0]
+        self.sync_blocks()
+
+        # Let the block reward mature and send coins including both
+        # the existing balance and the block reward.
+        self.nodes[0].generate(150)
+        assert_equal(self.nodes[1].getbalance(), 10 + 25)
+        txid = self.nodes[1].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=30, subtractfeefromamount=True)
+
+        # Orphan the block reward and make sure that the original coins
+        # from the wallet can still be spent.
+        self.nodes[0].invalidateblock(blk)
+        self.nodes[0].generate(200)
```

Comment:
I wasn't sure why this `generate(200)` was needed, so I commented it out, and the test hung in the next line (`sync_blocks()`). I restored the line and played around with different generate counts, and found that it fails with 151 and works at 152. If I'm counting correctly, we invalidated 151 blocks, so it seems that it's necessary to extend the chain so it's _longer_ than it was before. (Necessary for this test framework, not in real life.)

You could reduce this to 152; I like tests to "just barely" pass because it seems like that imparts better understanding to readers (I wondered is, "why is this 200?"), and also because if the code is acting differently from our understanding, for example, we think generating 152 blocks should work but it actually takes 170, and the test does 200, then that mistaken understanding (or even a bug in the code) would be papered over. 

Also, the test runs a little faster if we reduce this.

Of course, a downside to tests "just barely" passing is that they're more fragile. It's painful when someone makes a perfectly good change and one of the tests break! (I don't think there's much of a chance of that happening here.)

Reply:
Sounds good to me, updated to 152.